### DISCUSSION TITLE
Developing a Plugin for BabylonJS and need ideas to make it work

### DISCUSSION THREAD

**Discussion by Alecell** [webpack/webpack#16684] [Q&A]
**TL;DR**
How can I change a typescript module in memory level and return it to get processed by webpack?

----

Hey all, I'm trying to create a plugin for webpack to work longside with BabylonJS. The idea is simple, allow the user to import the model from the current folder, instead of add it globally on a asset folder, just like how we do with SCSS for React for example.

My idea was "simple", for each module of the project, get its content as string, search for calls of `SceneLoader` on it and for each call get the path of it (starting for the current folder of the module). With that path, get the model content, get the textures that the model imports and copy all of them (both model and texture) in the dist folder inside a folder with a hashed name (`dist/modules/[hash]/`). Also I want to repath the `SceneLoader` to that new path on dist, only on build/dev-server, dont changing the actuall file, changing it only on memory level for webpack process.

This is being a HUGE challenge for me cause I've never worked with webpack plugins and everything I tried didn't work, the cloeses I got to what I wanted was using the `compiler.hooks.compilation` and `compilation.hooks.buildModule` filtering for files that ends with `.ts`, but when I try to access the `_source` to get the file content its not exactly the module before it got processed (it has some changes compared to the original file) but eitherway I cannot access the file content because it is `null` leading to a error. 

I also thought to create a webpack loader, but for it I would need to communicate between the loader and plugin (since loader is just to proccess the files) and I wasnt able to make such communication work also.

Its also important to note that I'm not able to `import` the modules with the `import` keyword as the way we import models on BabylonJS is just a plain method such as `SceneLoader.ImportMeshAsync()`.

I have several questions about this, so:
- Is there a way to do what I wanted with webpack?
- What would be the right way to make something like this?
- Am I tryign to solve it in a "dumb way"? What would be a better solution?
